src/models/ModelBuilder.cc

```diff
@@ -3,23 +3,136 @@
 //
 
 #include "ModelBuilder.h"
+#include <unordered_set>
 
-void ModelBuilder::processSubstitutions(Logic::SubstMap const & subst) {
-    Logic::SubstMap copy;
-    subst.copyTo(copy);
-    logic.substitutionsTransitiveClosure(copy);
-    auto assignCopy = this->assignment;
-    auto model = this->build();
-    for (auto key : copy.getKeys()) {
-        if (logic.isIte(key)) {
-            // We store only values of reals variables
+std::unique_ptr<Model> ModelBuilder::build() {
+    std::unordered_map<SymRef,TemplateFunction,SymRefHash> builtDefinitions;
+    for (auto & symbolSigVal : definitions) {
+        const ValuationNode * valuationNode = symbolSigVal.second.second;
+        SRef sr = logic.getSortRef(symbolSigVal.first);
+        PTRef body = valuationTreeToFunctionBody(valuationNode, sr);
+        TemplateFunction templateFun(std::move(symbolSigVal.second.first), body);
+        builtDefinitions[symbolSigVal.first] = std::move(templateFun);
+    }
+    return std::unique_ptr<Model>(new Model(logic, assignment, builtDefinitions));
+}
+
+void ModelBuilder::addToTheoryFunction(SymRef sr, vec<PTRef> vals, PTRef val)
+{
+    if (logic.getSortRef(sr) != logic.getSortRef(val)) {
+        throw OsmtApiException("Incompatible sort for symbol valuation");
+    }
+    Symbol & sym = logic.getSym(sr);
```

Comment:
```suggestion
    Symbol sym = logic.getSym(sr);
```

Reply:
The copy doesn't seem necessary.  Maybe you meant `const Symbol & sym = logic.getSym(sr)` ?

Reply:
Ah, sorry, I got `Symbol` confused with `SymRef`. You are right.